base/compiler/types.jl

```diff
@@ -189,3 +188,14 @@ lock_mi_inference(ni::NativeInterpreter, mi::MethodInstance) = (mi.inInference =
     See lock_mi_inference
 """
 unlock_mi_inference(ni::NativeInterpreter, mi::MethodInstance) = (mi.inInference = false; nothing)
+
+"""
+Emit an annotation when inference widens a result. These annotations are ignored
+by the native interpreter, but can be used by external tooling to annotate
+inference results.
+"""
+mark_dynamic!(ni::NativeInterpreter, sv, s) = nothing
```

Comment:
This naming seems rather poor, since there's not really a sense in which the places this got called is related to "when inference widens a result". The only place that happens is `tmerge`. Perhaps this should be called simply `add_remark`?

Reply:
Fair enough. The idea here was that this answers the question "why did a dynamic dispatch get introduced here". I'm fine with just generically calling it `add_remark` though. Eventually we might want to have multiple classes of remarks, not just about inference precision, so that seems like a fine direction to go.

Reply:
Yeah, but the places we're putting it (in inference) aren't introducing dynamic dispatch (that's always there), they're merely decision points where we prune our analysis of the call graph.

Reply:
Depends on what you mean by dynamic dispatch ;). I meant calls to jl_apply_generic which is often what happens when we prune the analysis here. The reason to care about that is that that'd be what's disallowed in e.g. a GPU or XLA context. Anyway, add_remark! is fine.